crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -124,12 +128,34 @@ impl Trade {
         })
     }
 
+    fn adjusted_order_limits(&self) -> Result<(eth::U256, eth::U256), Error> {
+        match self.policies.first() {
+            Some(order::FeePolicy::PriceImprovement {
+                factor: _,
+                max_volume_factor: _,
+                quote,
+            }) => adjust_quote_to_order_limits(
+                self.sell.amount.0,
+                self.buy.amount.0,
+                self.side,
+                quote.sell.amount.0,
+                quote.buy.amount.0,
+                quote.fee.amount.0,
+            )
+            .map_err(|e| e.into()),
+            _ => Ok((self.sell.amount.0, self.buy.amount.0)),
+        }
+    }
+
     /// Surplus based on custom clearing prices returns the surplus after all
     /// fees have been applied.
     ///
     /// Denominated in NATIVE token
     fn native_surplus(&self, prices: &auction::Prices) -> Result<eth::Ether, Error> {
-        let surplus = self.surplus().ok_or(Error::Surplus(self.sell, self.buy))?;
+        let (limit_sell_amount, limit_buy_amount) = self.adjusted_order_limits()?;
```

Comment:
it's really subjective, but I don't like that we call this method unconditionally when really this is only needed for price improvement fee policies. Can you move it into the `PriceImprovement` branch please?

Reply:
Refactored a little bit. Not sure I got the idea, since `native_surplus` function doesn't care about the protocol fee, so there is no `PriceImprovement` branch for this function.